### ISSUE TITLE
expand('%:p') expands to $HOME when open file is named ~

### ISSUE THREAD

**Issue by azsn** [vim/vim#3072]
When the current file opened is named exactly `~` (that is, `expand('%')` would return `~`), `expand('%:p')` returns the expansion of `~` instead of the actual path to the current file. For example:

```
:cd /home/aidan/documents
:e ~
:echo expand('%:p')
```
This prints `/home/aidan` when `/home/aidan/documents/~` is expected.

I have tested this with no plugins or scripts on multiple versions of Vim, including 8.1, and on both Mac and Linux systems.

**Comment by tonymec** [comment]
I confirm this on gvim 8.1.105 on openSUSE Linux 15.0

`:e ~` or `:new ~` opens an empty window and says at the bottom `"~" [New File]` but then `:echo expand('%:p')` answers with the full path to my home directory, with a slash at the end.

To open a netrw window on my home directory I need to type `:e ~/` or `:new ~/` with a trailing slash. Vim then may or may not give something that looks like an error message but the netrw window displays my home directory contents anyway. Then `:echo expand('%:p')` displays the same value as when `~` was a file, i.e., the full path to my home directory with a trailing slash.

Maybe the `expand()` function cannot tell these two cases apart?


Best regards,
Tony.

**Comment by chrisbra** [comment]
Hm, this works:
` vim --clean -c ':echo expand("%:p")' ./documents/~` while this doesn't:
`vim --clean -c ':cd ~/documents' -c ':e ~'  -c ':echo expand("%:p")'`

**Comment by chrisbra** [comment]
Okay, this happens, because `expand('~')` expands your home directory. Expand
does not know, that the '~' came from a filename and should therefore be
expanded as a filename instead of a homedirectory.
You can use `:echo fnamemodify('./'.expand('%'), ':p')` to force evaluation as a file path.

Not sure how to fix this one.

**Comment by jamessan** [comment]
> Okay, this happens, because expand('~') expands your home directory.

But the user isn't calling `expand('~')`.  That's the result of handling the filename modifiers.  Shouldn't Vim be able to tell it doesn't need to re-evaluate expansions?

**Comment by azsn** [comment]
I've been looking at the source code; the function `eval_vars()` in ex_docmd.c is responsible for converting the `%` into a `~`, but then it passes that result into `modify_fname()` in eval.c which handles the `:p`. So `modify_fname()` sees `~:p`. However, `modify_fname()` also gets passed in the original string, so I'm thinking it may be possible for it to do a check to see if the original string was exactly `%:p`, and if so, treat the `~` as a filename instead of the home directory. A bit of a hack, but it might be the easiest way.

And thanks chrisbra for the workaround command; I was looking for something like that but my Vimscript skills weren't good enough.

**Comment by azsn** [comment]
I have created a patch which appears to fix the issue. I'm worried it might not work properly on Windows systems, but I don't have a good way to test that. Is a pull request the best way to submit the patch?

**Comment by chrisbra** [comment]
either way is fine. A Test would also be good. If you want to test a windows patch, you can fork the [vim-win32-installer](https://github.com/vim/vim-win32-installer) repository and throw the patch in the patch directory and have appveyor build it (this needs however a working setup for appveyor, which might be a bit tricky).

**Comment by tonymec** [comment]
@zelbrium : Did you test that in a netrw window on your home directory, expand('%:p') **should** return the full path to your home directory?

Best regards,
Tony.

**Comment by azsn** [comment]
Yes, I did. It seems that Vim, when opening your home directory (such as with `:e ~/`) automatically converts that to the directory's absolute path (`:echo expand('%')` returns the full path to the home directory) which means it still works correctly. I'm about to submit the patch to the vim-dev mailing list, if you would like to test it on your machine.

**Comment by azsn** [comment]
It seems my email to vim-dev@vim.org got lost. I'll post the patch here instead.

```diff
diff --git a/src/ex_docmd.c b/src/ex_docmd.c
index 4d5036f22..0602b82ef 100644
--- a/src/ex_docmd.c
+++ b/src/ex_docmd.c
@@ -10612,6 +10612,26 @@ find_cmdline_var(char_u *src, int *usedlen)
     return -1;
 }
 
+/*
+ * Issue #3072
+ * When eval_vars() is used on a string of '%:p' and the current
+ * filename is exactly '~', the filename is expanded to the home
+ * directory instead of the full path to the file. This replaces
+ * the '~' name with './~' so the path expansion works correctly.
+ */
+static void fix_tilde_filename(char_u *src, int usedlen, char_u **result, char_u **resultbuf) 
+{
+    if ((*result)[0] == '~'
+	&& (*result)[1] == NUL
+	&& src[usedlen] == ':'
+	&& src[usedlen + 1] == 'p')
+    {
+	*result = vim_strnsave(*result, 4);
+	*resultbuf = *result; /* remember allocated string */
+	vim_strncpy(*result, (char_u *)"./~", 4);
+    }
+}
+
 /*
  * Evaluate cmdline variables.
  *
@@ -10715,7 +10735,10 @@ eval_vars(
 		    valid = 0;	    /* Must have ":p:h" to be valid */
 		}
 		else
+		{
 		    result = curbuf->b_fname;
+		    fix_tilde_filename(src, *usedlen, &result, &resultbuf);
+		}
 		break;
 
 	case SPEC_HASH:		/* '#' or "#99": alternate file */
@@ -10779,7 +10802,10 @@ eval_vars(
 			valid = 0;	    /* Must have ":p:h" to be valid */
 		    }
 		    else
+		    {
 			result = buf->b_fname;
+			fix_tilde_filename(src, *usedlen, &result, &resultbuf);
+		    }
 		}
 		break;
 
diff --git a/src/testdir/test_expand.vim b/src/testdir/test_expand.vim
index c099edae1..0161f7cb3 100644
--- a/src/testdir/test_expand.vim
+++ b/src/testdir/test_expand.vim
@@ -39,3 +39,9 @@ func Test_with_tilde()
   call delete('Xdir ~ dir', 'd')
   call assert_false(isdirectory('Xdir ~ dir'))
 endfunc
+
+" Issue #3072
+func Test_expand_tilde_filename()
+  e ~
+  call assert_notequal(expand('%:p'), expand('~/'))
+endfunc

```

**Comment by tonymec** [comment]
AFAICT, it works.

Tested in gvim 8.1.115 (Huge) as modified above for Linux64, in windows on all three of:
- a file named `~`
- a netrw buffer on my home directory
- some other file with a "normal" name.

The corresponding Tiny version compiles successfully. No other test possible there since `expand()` is undefined because Tiny builds are compiled with `-eval`.

I did not run the provided test.

Best regards,
Tony.
